File: ci/src/freecam_ci/dev.py
import os
import subprocess
import sys

from .project_files import CI

ANSI_ESC = "\x1b"
ANSI_COLOR_RESET = ANSI_ESC + "[39m"
ANSI_RED = ANSI_ESC + "[31m"
ANSI_BOLD = ANSI_ESC + "[1m"
ANSI_BOLD_RESET = ANSI_ESC + "[22m"
GHA = os.getenv("GITHUB_ACTIONS") == "true"


def error(msg: str):
    if GHA:
        msg = "::error::" + msg
    elif sys.stderr.isatty():
        msg = ANSI_BOLD + ANSI_RED + msg + ANSI_COLOR_RESET + ANSI_BOLD_RESET
    print(msg, file=sys.stderr)


def run(*cmd: str) -> int:
    print("+", " ".join(cmd), file=sys.stderr)
    result = subprocess.run(cmd)
    if result.returncode != 0:
        error(f"FAILED: exit code {result.returncode}")
    return result.returncode


def run_all(*cmds: list[str]) -> None:
    results = [run(*cmd) for cmd in cmds]
    errors = sum(result != 0 for result in results)
    if errors > 0:
        s = "s" if errors > 1 else ""
        error(f"{errors} error{s} occurred")
        sys.exit(1)


def check_python() -> None:
    run_all(
        ["ruff", "format", "--check"],
        ["ruff", "check"],
        ["mypy", str(CI)],
        ["pytest"],
    )


def format_python() -> None:
    status = run("ruff", "format")
    sys.exit(status)

File: ci/src/dev.ts
import { spawnSync } from "node:child_process";

const GHA = process.env.GITHUB_ACTIONS === "true";

function error(msg: string) {
  if (GHA) console.error(`::error::${msg}`);
  else console.error(msg);
}

function run(cmd: string[]): number {
  console.error("+", cmd.join(" "));
  const arg0 = cmd[0];
  if (!arg0) return 0;

  const result = spawnSync(arg0, cmd.slice(1), { stdio: "inherit" });
  const code = result.status ?? 1;

  if (code !== 0) error(`FAILED: exit code ${code}`);
  return code;
}

function runAll(cmds: string[][]) {
  const results = cmds.map(run);
  const failures = results.filter((c) => c !== 0).length;

  if (failures > 0) {
    error(`${failures} failure${failures > 1 ? "s" : ""} occurred`);
    process.exit(1);
  }
}

if (import.meta.main) {
  runAll([
    ["npm", "run", "lint"],
    ["npm", "run", "typecheck"],
    ["node", "--test"],
  ]);
}
